django/db/models/base.py

```diff
@@ -630,6 +630,18 @@ def save(self, force_insert=False, force_update=False, using=None,
         that the "save" must be an SQL insert or update (or equivalent for
         non-SQL backends), respectively. Normally, they should not be set.
         """
+        for field in self._meta.concrete_fields:
+            if field.is_relation:
+                try:
+                    obj = getattr(self, field.name, None)
```

Comment:
Isn't there a risk of this doing a database query? That doesn't seem acceptable.

Can't we first check the plain FK attribute (i.e. `field_id`), and only check further if it is `None`? If it's not `None` then clearly the assigned object had an ID. If it is `None` then we just need to see if there's an unsaved object in the "main" attribute or not.